### ISSUE TITLE
[feature] More precise “modes” for data modified by a subprogram

### ISSUE THREAD

**Issue by clairedross** [AdaCore/ada-spark-rfcs#126]
### Summary

A new contract could be used to precisely express which parts of a parameter/global of mode IN OUT are modified/preserved by a subprogram. It would be interesting to also allow expressing in which cases these (parts of) objects are preserved. 

```ada
   type R is record
      F1, F2 : Boolean;
   end record;

   procedure Write_F1 (X : in out R) with
     Modifies => (others => X.F1);

   procedure Write_F1 (X : in out R) is
   begin
      X.F1 := False;
   end Write_F1;

   G : Integer;

   procedure Write_G_If_B (B : Boolean) with
     Global => (In_Out => G),
     Modifies => (B => G);

   procedure Write_G_If_B (B : Boolean) is
   begin
      if B then
         G := 0;
      end if;
   end Write_G_If_B;
```
More generally, this contract could be structured like a Contract_Cases with various cases made of a precondition and a sequence of paths into parameters/globals of mode IN OUT. Paths here could be defined like for deep delta aggregates.

### Motivation

Modes for parameters and global variables modified by a subprogram are sometimes too coarse, it would be useful to easily express which parts of the data are preserved or modified in a contract.

### Caveats and alternatives

This contract is theoretically equivalent to a contract cases where the consequences could be written as `X = (X'Old with delta Path => X.Path'Old ...)`. It can be handled that way for proof, but it is more complex for execution as the equality here is not the Ada equality but the logical one, so this contract would probably remain non-executable (like Global and Depends).

An alternative would be to rather list preserved parts using a `Preserves` annotation.

**Comment by clairedross** [comment]
> "Modifies" seems more maintainable than "Preserves" in that the number of "Modifies" correlates with the number of updates performed, while "Preserved" goes just the opposite. With "Preserved", the moment you update a single field, you need to specify the full list of other fields in a "Preserves", which seems burdensome and probably too noisy.

I agree with that. I also like "Modifies" better for tagged types as hidden components are preserved by default.

> I do wonder what you expect to happen with a private type. Ada 2022 has the "Stable_Properties" aspect which is more like Preserved, but is generally at the type level, but can be overridden at an individual subprogram level (see [RM 7.3.4](https://ada-rapporteur-group.github.io/ARM/Ada_202Y/AA-7-3-4.html)). It seems worth considering something like that for SPARK for private types.

I agree that private types are an issue. We already have a workaround though, the "equivalent" formulation listed in the alternatives can be used inside an expression function and in a contract.

**Comment by raph-amiard** [comment]
Is the plan to implement this as a runtime contract too ?

 It looks like the expansion to the contract_case would be awfully slow to check as a runtime check, and almost prohibitive to check in a majority of cases.

Appart from that, no comment really :slightly_smiling_face: This looks like an information that is really only useful for proof, and that you would never use in pure Ada.

**Comment by clairedross** [comment]
> Is the plan to implement this as a runtime contract too ?

As stated in the Discussion, I am not sure it is possible to check at runtime due to the fact that it requires logical equality. Appart from that, I am not sure why you think it is more prohibitive than other contracts.

**Comment by florianschanda** [comment]
I think a very important question here if the following is valid:
```
type Rec is record
   A : Integer;
   B : Integer;
   C : Integer;
end record;

procedure P (X : in out Rec)
with Modifies => (others => X.C)
is
begin
   X.B := 12;
   X.C := 12;
end Modifies;
```

I.e. Is 
* `Modifies` a positive statement to modify some value;
* or is it ALSO a statement to not touch anything else

I would hope/assume that it's the second, but from reading the summary it was not entirely obvious to me.

Assuming it's the second, there is another question open:

```
procedure P (X : in out Rec)
with Modifies => (others => X.C)
is
begin
   X.C := X.A;
end Modifies;
```

* When calling P, can flow positively conclude that X.C does not depend on X.A?
* Is the body of P of even legal?

It would perhaps be nice to also support a different syntax that just allows you to enumerate modified things without detailing how:

```
procedure P (X : in out Rec)
with Modifies => (X.A, X.C)
--  Implicitly means Post => X.B = X'Old.B
```

**Comment by clairedross** [comment]
> I.e. Is
> 
> Modifies a positive statement to modify some value;
> or is it ALSO a statement to not touch anything else

In fact it is not even an ALSO. Modifies means that everything which is not mentioned is preserved. The rest might be modified or not. Note also that what is not mentioned should be preserved, not "not modified" as it is fine to modify it, it should just be reset to the same value before the subprogram returns. As mentioned in the RFC, it is more or less syntactic sugar for Post => `X = (X'Old with delta Path => X.Path ...)`, where "=" is the logical equality. So it is a proof annotation, not a flow one.

> procedure P (X : in out Rec)
> with Modifies => (others => X.C)
> is
> begin
>    X.C := X.A;
> end Modifies;
> When calling P, can flow positively conclude that X.C does not depend on X.A?
> Is the body of P of even legal?

The body of P is legal, and there is no reason for flow analysis to conclude anything from this (proof) contract. Maybe I don't understand the question. Note that the arrows are not dependencies, they are guards, like in contract cases.

> It would perhaps be nice to also support a different syntax that just allows you to enumerate modified things without detailing how.

You mean without any guards. Sure, it could be an enhancement to make the "others" guard optional.

**Comment by florianschanda** [comment]
Oh I really mis-understood the LHS of the modifies items, I thought it's basically a depends contract.

OK then I am fully on board with everything here, but having the short-hand where just a list of items is the same as `others => (...` would be really good for usability.

**Comment by florianschanda** [comment]
Two more questions:
* how will this interact with `Abstract_State`
* how will this interact with `limited private`

**Comment by clairedross** [comment]
Very interesting questions. Currently it does not, though we could imagine a Refined_Modifies for internal use later. Do you often use Refined_Post? Were you planning on using your enhancement of delta aggregates in such a post? Similarly, I have not proposed Modifies'Class for now...

**Comment by florianschanda** [comment]
I think Refined_Post is pretty rare in practice. It would be good to have eventual parity with Post, but it doesn't feel super urgent.

I have a strong preference for limited private / state working, but perhaps we shouldn't look for "equality" but instead mean "I didn't touch it". Right now you can't express this property through a post-condition, which is a long-standing issue that I have (since 2014) :)

This feature is way more interesting / applicable for proof anyway, I don't think it is terribly important that it's executable.

**Comment by clairedross** [comment]
> I have a strong preference for limited private / state working, but perhaps we shouldn't look for "equality" but instead mean "I didn't touch it". Right now you can't express this property through a post-condition, which is a long-standing issue that I have (since 2014) :)

I think I misunderstood your question. I think abstract states, as well as private and limited object should be allowed to occur in Modifies (or to not occur...)
